File: dynamic_listings/C++/default/maths/probability_luck.cpp
#include <algorithm>

enum Rarity {EPIC, RARE, UNCOMMON, COMMON};
// Our probabilities, from least to most common
int pool[4][2] = {
    {EPIC, 5},
    {RARE, 15},
    {UNCOMMON, 30},
    {COMMON, 50},
};

// Our "luck stat": each point gives 1% more chance to get a higher-tier item
int LUCK = 25;

// We cap the Luck stat at 100, the limit is 100% epic items
LUCK = std::min(LUCK, 100);

// We "overload" the prize pool, making the sum go over 100%
int overloaded_pool[4][2];
int overload_factor = 0;

for (int i = 0; i < 4; i++){
    int new_prob = pool[i][1] + LUCK;
    // We accumulate the overload factor for further calculation
    overload_factor = overload_factor + new_prob;
    overloaded_pool[i][0] = pool[i][0];
    overloaded_pool[i][1] = new_prob;
}

// We calculate how much we "overloaded" the prize pool
overload_factor = overload_factor - 100;

// We rebalance the prizes to a total of 100, from most to least common
int rebalanced_pool[4][2];
// We need to start from the most common, which means we will iterate backwards
for (int i = 3; i >= 0; i--){
    const int item = overloaded_pool[i][0];
    const int probability = overloaded_pool[i][1];
    // This will be modified later, if the pool is "overloaded"
    int new_probability = probability;
    // If the prize pool is still "overloaded"
    if (overload_factor > 0){
        // We calculate a "discharge factor" of sorts
        int value_to_remove = std::min(probability, overload_factor);
        // We reduce our "overload"
        overload_factor = overload_factor - value_to_remove;
        // And put the new probability for the class
        new_probability = probability - value_to_remove;
    }
    // We append the new pool item
    rebalanced_pool[i][0] = item;
    rebalanced_pool[i][1] = new_probability;
}

File: dynamic_listings/rust/default/maths/probability_luck.rs
#[derive(Clone, Copy)]
enum Rarity {
    Common,
    Uncommon,
    Rare,
    Epic,
}

// Our probabilities, from least to most common
const POOL: [(Rarity, u8); 4] = [
    (Rarity::Epic, 5),
    (Rarity::Rare, 15),
    (Rarity::Uncommon, 30),
    (Rarity::Common, 50),
];

# fn main() {
// Our "luck stat": each point gives 1% more chance to get a higher-tier item
let mut luck = 25;

// We cap the Luck stat at 100, the limit is 100% epic items
luck = u8::min(luck, 100);

// We "overload" the prize pool, making the sum go over 100%
let mut overloaded_pool: [(Rarity, u8); 4] = POOL;
let mut overload_factor = 0;

for x in 0..4 {
    let new_prob = POOL[x].1 + luck;
    // We accumulate the overload factor for further calculation
    overload_factor += new_prob;
    overloaded_pool[x].0 = POOL[x].0;
    overloaded_pool[x].1 = new_prob;
}

// We calculate how much we "overloaded" the prize pool
overload_factor = overload_factor - 100;

// We rebalance the prizes to a total of 100, from most to least common
let mut rebalanced_pool: [(Rarity, u8); 4] = POOL;
// We need to start from the most common, which means we will iterate backwards
for x in (0..=3).rev() {
    let item = overloaded_pool[x].0;
    let probability = overloaded_pool[x].1;
    // This will be modified later, if the pool is "overloaded"
    let mut new_probability = probability;
    // If the prize pool is still "overloaded"
    if overload_factor > 0 {
        // We calculate a "discharge factor" of sorts
        let value_to_remove = u8::min(probability, overload_factor);
        // We reduce our "overload"
        overload_factor = overload_factor - value_to_remove;
        // And put the new probability for the class
        new_probability = probability - value_to_remove;
    }
    // We append the new pool item
    rebalanced_pool[x].0 = item;
    rebalanced_pool[x].1 = new_probability;
}
# }
